lib/granite/context_proxy.rb

```diff
@@ -0,0 +1,39 @@
+require 'granite/context_proxy/data'
+require 'granite/context_proxy/proxy'
+
+module Granite
+  # This concern contains class methods used for actions and projectors
+  #
+  module ContextProxy
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      def using(data)
+        Proxy.new(self, Data.wrap(data))
+      end
+
+      def as(performer)
+        using(performer: performer)
+      end
+
+      def with_context(context)
+        key = proxy_context_key
+        old_context = Thread.current[key]
+        Thread.current[key] = context
+        yield
+      ensure
+        Thread.current[key] = old_context
```

Comment:
Is this not causing a memory leak @andriusch ? At not point we delete the key from this storage. And the key is dynamic `:"granite_proxy_performer_#{hash}"`. So as long as this thread is processing requests, we will be adding more and more keys there. They might in the end be pointing to `nil` but still.

Reply:
It's not causing a leak, just bloat, since we're taking `BA.hash` not `BA.new.hash`. I'd love to keep it localized, but the only way I can come up with is this:
```ruby
def using(data)
  ctx = Data.wrap(data)
  Class.new(self) do
    define_method(:ctx) { ctx }
  end
end
```

And I'm on the fence whether creating classes like that is a good thing.

Reply:
> And I'm on the fence whether creating classes like that is a good thing.

Point for creativity in coming up with that dynamic class idea but it would be worse than what what we have now. I would be also afraid it could trash some ruby internal cache optimizations.

Reply:
> we're taking BA.hash

Aha, indeed. I missed this is in `ClassMethods`